client/app/use/useSave.ts

```diff
@@ -1,37 +1,46 @@
-import Vue from 'vue';
-import { ref, computed } from '@vue/composition-api';
+import { ref } from '@vue/composition-api';
 import Track, { TrackId } from 'vue-media-annotator/track';
 
 import { saveDetections } from 'app/api/viameDetection.service';
 import { setMetadataForFolder } from 'app/api/viame.service';
 
-export default function useSave() {
-  const pendingSaveCountMap = ref({ default: 0 } as Record<string, number>);
-
-  const pendingSaveCount = computed(() => Object.values(pendingSaveCountMap.value)
-    .reduce((p, c) => p + c, 0));
+export default function useSave(datasetId: string) {
+  const pendingSaveCount = ref(0);
+  const pendingChangeMap = {
+    upsert: new Map<TrackId, Track>(),
+    delete: new Set<TrackId>(),
+  };
 
   async function save(
-    datasetId: string,
-    trackMap?: Map<TrackId, Track>,
     datasetMeta?: object,
   ) {
-    if (trackMap && pendingSaveCountMap.value.default) {
-      await saveDetections(datasetId, trackMap);
-      Vue.set(pendingSaveCountMap.value, 'default', 0);
+    if (pendingChangeMap.upsert.size || pendingChangeMap.delete.size) {
+      await saveDetections(datasetId, {
+        upsert: pendingChangeMap.upsert,
+        delete: Array.from(pendingChangeMap.delete),
+      });
+      pendingChangeMap.upsert.clear();
+      pendingChangeMap.delete.clear();
     }
-    if (datasetMeta && pendingSaveCountMap.value.meta) {
+    if (datasetMeta) {
       await setMetadataForFolder(datasetId, datasetMeta);
-      Vue.set(pendingSaveCountMap.value, 'meta', 0);
     }
+    pendingSaveCount.value = 0;
   }
 
-  function markChangesPending(name = 'default') {
-    if (name in pendingSaveCountMap.value) {
-      pendingSaveCountMap.value[name] += 1;
-    } else {
-      Vue.set(pendingSaveCountMap.value, name, 1);
+  function markChangesPending(
+    type: 'upsert' | 'delete' | 'other' = 'other',
+    track?: Track,
+  ) {
+    if (type === 'delete' && track !== undefined) {
+      pendingChangeMap.delete.add(track.trackId);
+    } else if (type === 'upsert' && track !== undefined) {
+      pendingChangeMap.delete.delete(track.trackId);
+      pendingChangeMap.upsert.set(track.trackId, track);
+    } else if (type !== 'other') {
+      throw new Error('Arguments inconsistent with pending change type');
```

Comment:
I think you may want to remove from the upsert Set when you delete an object as well.  This relates to another line of code within the python stuff that has you deleting objects before inserting them.  
Meaning with a blank slate:
- create track 0 and edit
- create track 1 and edit
- create track 2 and edit
- delete track 1

now your delete has a deletion for track 1, but it also has an upsert for track 1 which will cause errors when saving

Reply:
You're right, I missed an edge case.

Reply:
Resolved in 349c5c5